plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -280,21 +287,24 @@ we opted for option 1, which also happens to be simpler to implement.
 -}
 
 {- Note [Structure of the step counter]
-The step counter is kept in a 'WordArray', which is 8 'Word8's packed into a single 'Word64'.
-This happens to suit our purposes exactly, as we want to keep a counter for each kind of step
-that we know about (of which there are 7) and one for the total number.
+The step counter is kept in a 'MutablePrimArray', which is a fast way of storing a bunch of
+mutable 'Word8's.
+This suits our purposes, as we need one counter for every step type, and one for the total number.
 
-We keep the counters for each step in the first 7 indices, so we can index them simply by using
+We keep the counters for each step in the first indices, so we can index them simply by using
 the 'Enum' instance of 'StepKind', and the total counter in the last index.
-
-Why use a 'WordArray'? It optimizes well, since GHC can often do quite a lot of constant-folding
-on the bitwise operations that get emitted. We are restricted to counters of size 'Word8', but this
-is fine since we will reset when we get to 200 steps.
-
-The sharp-eyed reader might notice that the benchmarks in 'word-array' show that 'PrimArray'
-seems to be faster! However, we tried that and it was slower overall, we don't know why.
 -}
 
+-- | The number of step counters that we need, should be the same as the number of constructors
+-- of 'StepKind'.
+type NumberOfStepCounters = 7
```

Comment:
If you take out `StepKind` into a separate module to alleviate GHC stage restriction you could enforce it at compile-time by:

```suggestion
type NumberOfStepCounters = $[t| $(pure $ LitT $ NumTyLit $ fromIntegral $ 1 + fromEnum (maxBound @StepKind) ) |]```

Reply:
> alleviate GHC stage restriction you could enforce it at compile-time by:

You can also do that by inserting

```haskell
$(return [])
```

we do it [here](https://github.com/input-output-hk/plutus/blob/e714ddabe7904a64fe05ea59f158106cb6f303d8/plutus-core/plutus-core/src/Universe/Core.hs#L727).